validator/sawtooth_validator/gossip/structure_verifier.py

```diff
@@ -0,0 +1,153 @@
+from google.protobuf.message import DecodeError
+
+from sawtooth_validator.protobuf import client_pb2
+from sawtooth_validator.protobuf.batch_pb2 import BatchHeader
+from sawtooth_validator.protobuf.batch_pb2 import Batch
+from sawtooth_validator.protobuf.block_pb2 import Block
+from sawtooth_validator.protobuf import network_pb2
+from sawtooth_validator.protobuf.network_pb2 import GossipBlockResponse
+from sawtooth_validator.protobuf.network_pb2 import GossipBatchResponse
+from sawtooth_validator.networking.dispatch import Handler
+from sawtooth_validator.networking.dispatch import HandlerResult
+from sawtooth_validator.networking.dispatch import HandlerStatus
+from sawtooth_validator.protobuf.validator_pb2 import Message
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+def is_valid_block(block):
+    if not all(map(is_valid_batch, block.batches)):
+        return False
+
+    return True
+
+
+def is_valid_batch(batch):
+    # batch structure verification
+    header = BatchHeader()
+    header.ParseFromString(batch.header)
+
+    # validate the transaction_ids field in batch header contains a list of
+    # transaction header_signatures and must be the same order as the
+    # transactions field
+    if len(batch.transactions) > len(header.transaction_ids):
+        LOGGER.debug("Batch has extra transactions. Dropping batch: %s",
+                     batch.header_signature)
+        return False
+    elif len(batch.transactions) < len(header.transaction_ids):
+        LOGGER.debug("Batch lacks transactions. Dropping batch: %s",
+                     batch.header_signature)
+        return False
+
+    for header_txn_id, txn in zip(header.transaction_ids, batch.transactions):
+        if header_txn_id != txn.header_signature:
+            LOGGER.debug("The header.transaction_ids does not match the "
+                         "order of transactions in the batch: %s txn: %s",
+                         batch.header_signature, txn.header_signature)
+            return False
+
+    return True
+
+
+class GossipHandlerStructureVerifier(Handler):
+    def handle(self, connection_id, message_content):
+        gossip_message = network_pb2.GossipMessage()
+        gossip_message.ParseFromString(message_content)
+        if gossip_message.content_type == "BLOCK":
+            block = Block()
+            block.ParseFromString(gossip_message.content)
+            if is_valid_block(block):
+                LOGGER.debug("block passes batch structure verification %s",
+                             block.header_signature)
+                return HandlerResult(status=HandlerStatus.PASS)
+
+            LOGGER.debug("block's batches structure is invalid: %s",
+                         block.header_signature)
+            return HandlerResult(status=HandlerStatus.DROP)
+        elif gossip_message.content_type == "BATCH":
+            batch = Batch()
+            batch.ParseFromString(gossip_message.content)
+            if not is_valid_batch(batch):
```

Comment:
This is going to be a really annoyingly small nit, and you can ignore, but above you check `if is_valid_block(...)` and here you check `if not is_valid_batch(...)`.  From a correctness standpoint, this is not a problem, but from a consistency standpoint, I think it would be better (not just for my anal-retentive ways, but for others looking at the code) if both conditionals either checked for the "success" case or the "failure" case.  If you choose to address this, I don't have much preference either way, except that the handlers below seems to look for the "failure" case in the conditional